### PR TITLE
add encoding/decoding of hmpb ballot page metadata

### PR DESCRIPTION
**Pull request by benadida**

### CODE REVIEW

File: src/v1/index.ts [hunk 1]
```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
...
   return votes
 }
+
+interface HMPBBallotPageMetadata {
+  election: Election
+  electionHash: string // a hexadecimal string
+  precinctId: string
+  ballotStyleId: string
+  locales: BallotLocale
+  pageNum: number
```
Comment (eventualbuddha) [inline src/v1/index.ts]: `pageNumber`?

File: src/v1/index.ts [hunk 2]
```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
...
+  isLiveMode: boolean
```
Comment (eventualbuddha) [inline src/v1/index.ts]: `isTestMode`? I know we have both, but I think this is more common.

File: src/v1/index.ts [hunk 3]
```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
...
+  isAbsenteeMode: boolean
```
Comment (eventualbuddha) [inline src/v1/index.ts]: Use `BallotType`?

File: src/v1/index.ts [hunk 4]
```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
...
+  const precincts = election.precincts
+  const ballotStyles = election.ballotStyles
+
+  const precinctIndex = precincts.findIndex((p) => p.id === precinctId)
+  const ballotStyleIndex = ballotStyles.findIndex(
+    (bs) => bs.id === ballotStyleId
+  )
+  const primaryLocaleIndex = SUPPORTED_LOCALES.indexOf(locales.primary)
+  const secondaryLocaleIndex = locales.secondary
+    ? SUPPORTED_LOCALES.indexOf(locales.secondary)
```
Comment (eventualbuddha) [inline src/v1/index.ts:391]: This will return -1 if it isn't a known locale. I think it should throw. Same with primary.

File: src/v1/index.ts [hunk 5]
```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
...
+    .writeUint(precinctIndex, { max: precincts.length })
+    .writeUint(ballotStyleIndex, { max: ballotStyles.length })
+    .writeUint(primaryLocaleIndex, { max: SUPPORTED_LOCALES.length })
+    .writeBoolean(!!secondaryLocaleIndex)
+
+  if (secondaryLocaleIndex) {
+    bits.writeUint(secondaryLocaleIndex, { max: SUPPORTED_LOCALES.length })
+  }
+
+  bits.writeUint(pageNum, { max: 30 })
```
Comment (eventualbuddha) [inline src/v1/index.ts]: Extract to a constant?

File: src/v1/index.ts [hunk 6]
```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
...
+  if (!bits.skipUint8(...HMPBPrelude)) {
+    throw new Error(
+      "expected leading prelude 'V' 'P' 0b00000001 but it was not found"
+    )
+  }
+
+  const precincts = election.precincts
+  const ballotStyles = election.ballotStyles
+
+  const electionHash = bits.readString({ encoding: HexEncoding })
```
Comment (eventualbuddha) [inline src/v1/index.ts:446]: Anything after this point could fail/crash if the hash is not as expected. Worth separating this somehow so an explicit message can be shown comparing the hashes?

File: src/v1/index.ts [hunk 7]
```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
...
+  isLiveMode,
+  isAbsenteeMode,
+  ballotId,
+}: HMPBBallotPageMetadata): Uint8Array {
+  const bits = new BitWriter()
+
+  const precincts = election.precincts
+  const ballotStyles = election.ballotStyles
+
+  const precinctIndex = precincts.findIndex((p) => p.id === precinctId)
```
Comment (eventualbuddha) [inline src/v1/index.ts:375]: Ensure these are actually found?

File: src/v1/index.ts [hunk 8]
```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
 
   return votes
 }
+
+interface HMPBBallotPageMetadata {
```
Comment (eventualbuddha) [inline src/v1/index.ts:347]: This is almost the same as `BallotPageMetadata` from hmpb-interpreter. Does it make sense to consolidate somehow?

**Review by eventualbuddha — APPROVED** [review]
Mostly looks good, though I have a few questions.

File: src/v1/index.ts (hunk 6 above)
Comment (benadida) [reply src/v1/index.ts:446]: I want to explicitly not crash / throw at this low level if there was a small text change that doesn't actually change the ballot layout, etc. I can imagine this happening by mistake. I'm thinking the hash comparison can happen at a higher level and we can choose to ignore it if we want to, enforce it if we prefer

File: src/v1/index.ts (hunk 8 above)
Comment (benadida) [reply src/v1/index.ts:347]: I'm thinking `hmpb-interpreter` should use this data structure in a future refactor?

Comment (eventualbuddha) [reply src/v1/index.ts:347]: Perhaps, but it'd have to be without `election` if this structure is to be serialized.

File: src/v1/index.ts (hunk 6 above)
Comment (eventualbuddha) [reply src/v1/index.ts:446]: What I mean is that the `read*` calls below may fail given a different enough election definition. If there is 1 precinct and 1 ballot style in one election, then 90 of both in another, that second one will use more bits to represent the precinct and ballot style indexes. Every call after that would be reading with an incorrect offset. If the padding or offset works out unfavorably, the final `readString` call in this function could trigger an EOF error.

Comment (benadida) [reply src/v1/index.ts:446]: Right. But then are you proposing that we hard fail if there is an election hash mismatch?

Comment (eventualbuddha) [reply src/v1/index.ts:446]: No, I'm suggesting it _will_ fail as written in some circumstances. Thus, if we want a reliable way to read the hash for comparison, we either need to separate reading the hash from reading the rest, or fix the size of the precinct and ballot style uints.

**Pull request merged**